tests/cpp_tests/test_common.cpp

```diff
@@ -0,0 +1,104 @@
+/*!
+ * Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE file in the project root for license information.
+ */
+#include <gtest/gtest.h>
+
+#include <limits>
+
+#include "../include/LightGBM/utils/common.h"
+
+
+class AtofPreciseTest : public testing::Test {
+ public:
+  struct AtofTestCase {
+    const char* data;
+    double expected;
+  };
+
+  static double TestAtofPrecise(
+      const char* data, double expected, bool test_eq = true) {
+    double got = 0;
+    const char* end = LightGBM::Common::AtofPrecise(data, &got);
+    EXPECT_TRUE(end != data) << "fail to parse: " << data;
+    EXPECT_EQ(*end, '\0') << "not parsing to end: " << data;
+    if (test_eq) {
+      EXPECT_EQ(expected, got) << "parse string: " << data;
+    }
+    return got;
+  }
+};
+
+TEST_F(AtofPreciseTest, Basic) {
+  AtofTestCase test_cases[] = {
+      { "0", 0.0 },
+      { "0E0", 0.0 },
+      { "-0E0", 0.0 },
+      { "-0", -0.0 },
+      { "1", 1.0 },
+      { "1E0", 1.0 },
+      { "-1", -1.0 },
+      { "-1E0", -1.0 },
+      { "123456.0", 123456.0 },
+      { "432E1", 432E1 },
+      { "1.2345678", 1.2345678 },
+      { "2.4414062E-4", 2.4414062E-4 },
+      { "3.0540412E5", 3.0540412E5 },
+      { "3.355445E7", 3.355445E7 },
+      { "1.1754944E-38", 1.1754944E-38 },
```

Comment:
Please add `FLT_MAX`, `DBL_MIN/MAX` and just a few corner cases, for instance [from here](https://github.com/Alexhuszagh/rust-lexical/blob/master/data/test-parse-unittests/strtod_tests.toml).